janus_server/src/aggregator.rs

```diff
@@ -439,11 +444,166 @@ where
         &self,
-        _task_id: TaskId,
-        _job_id: AggregationJobId,
-        _transitions: Vec<Transition>,
+        task_id: TaskId,
+        job_id: AggregationJobId,
+        transitions: Vec<Transition>,
     ) -> Result<AggregateResp, Error> {
-        todo!() // TODO(brandon): implement
+        let vdaf = Arc::new(self.vdaf.clone());
+        let verify_param = Arc::new(self.verify_param.clone());
+        let transitions = Arc::new(transitions);
+
+        // TODO(brandon): don't hold DB transaction open while computing VDAF updates?
+        // TODO(brandon): don't do O(n) network round-trips (where n is the number of transitions)
+        Ok(self
+            .datastore
+            .run_tx(|tx| {
+                let vdaf = vdaf.clone();
+                let verify_param = verify_param.clone();
+                let transitions = transitions.clone();
+
+                Box::pin(async move {
+                    // Read existing state.
+                    let (mut aggregation_job, report_aggregations) = future::try_join(
+                        tx.get_aggregation_job::<A>(task_id, job_id),
+                        tx.get_report_aggregations_for_aggregation_job::<A>(
+                            &verify_param,
+                            task_id,
+                            job_id,
+                        ),
+                    )
+                    .await?;
+
+                    // Handle each transition in the request.
+                    let mut report_aggregations = report_aggregations.into_iter();
+                    let (mut saw_continue, mut saw_finish) = (false, false);
+                    let mut response_transitions = Vec::new();
+                    for transition in transitions.iter() {
+                        // Match transition received from leader to stored report aggregation, and
+                        // extract the stored preparation step.
+                        let mut report_aggregation = loop {
+                            let mut report_agg = report_aggregations.next().ok_or_else(|| {
+                                warn!(?task_id, ?job_id, nonce = %transition.nonce, "Leader sent unexpected, duplicate, or out-of-order transitions");
+                                datastore::Error::User(Box::new(Error::UnrecognizedMessage(
+                                    "leader sent unexpected, duplicate, or out-of-order transitions",
+                                    task_id,
+                                )))
+                            })?;
+                            if report_agg.nonce != transition.nonce {
+                                // This report was omitted by the leader because of a prior failure.
+                                // Note that the report was dropped (if it's not already in an error
+                                // state) and continue.
+                                if !matches!(report_agg.state, ReportAggregationState::Waiting(_)) {
```

Comment:
I think this conditional is backwards, presently it will only overwrite failed/invalid/finished states with failed.